apollo-gradle-plugin-kotlin/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloPlugin.kt

```diff
@@ -0,0 +1,243 @@
+package com.apollographql.apollo.gradle.kotlin
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.tasks.factory.dependsOn
+import com.apollographql.apollo.compiler.GraphQLCompiler
+import org.gradle.api.Action
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.internal.HasConvention
+import org.gradle.api.internal.file.FileResolver
+import java.io.File
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import javax.inject.Inject
+import kotlin.IllegalArgumentException
+
+
+class ApolloPlugin : Plugin<Project> {
+  companion object {
+    const val TASK_GROUP = "apollo"
+  }
+
+  lateinit var apolloExtension: ApolloExtension
+
+  override fun apply(project: Project) {
+    apolloExtension = project.extensions.create("apollo", ApolloExtension::class.java)
+
+    // the extension block has not been evaluated yet, register a callback once the project has been evaluated
+    project.afterEvaluate {
+      afterEvaluate(it)
+    }
+  }
+
+  private fun afterEvaluate(project: Project) {
+    val generateClassesTask = project.tasks.register("generateApolloClasses") {
+      it.group = TASK_GROUP
+    }
+
+    val androidExtension = project.extensions.findByName("android")
+
+    apolloExtension.validate()
+
+    when {
+      androidExtension is LibraryExtension -> {
+        androidExtension.libraryVariants.all(Action { variant ->
+          registerAndroid(project, apolloExtension, variant, generateClassesTask)
+        })
+        androidExtension.testVariants.all(Action { variant ->
+          registerAndroid(project, apolloExtension, variant, generateClassesTask)
+        })
+      }
+      androidExtension is AppExtension -> {
+        androidExtension.applicationVariants.all(Action { variant ->
+          registerAndroid(project, apolloExtension, variant, generateClassesTask)
+        })
+      }
+      androidExtension == null -> {
+        val javaPlugin = project.convention.getPlugin(JavaPluginConvention::class.java)
+        val sourceSets = javaPlugin.sourceSets
+
+        sourceSets.forEach {
+          val apolloVariant = ApolloVariant(
+              name = it.name,
+              sourceSetNames = listOf(it.name)
+          )
+
+          val task = registerVariantTask(project, apolloExtension, apolloVariant) { task ->
+            if (!apolloExtension.generateKotlinModels) {
```

Comment:
I believe if this is a plugin `apollo-gradle-plugin-kotlin` for Kotlin projects this code gen flag must be always true?

Reply:
I'm thinking we could use the same plugin for kotlin and java, which is why I used the generic `com.apollographql.apollo` as a plugin id. The artifact name is suffixed by `-kotin` because it's written in kotlin but it should be applicable to other projects.

Reply:
Ultimately, we could have the plugin make `generateKotlinModels=true` by default if a kotlin plugin is detected.